haystack/components/preprocessors/document_splitter.py

```diff
@@ -132,36 +132,80 @@ def _concatenate_units(
 
         text_splits: List[str] = []
         splits_pages = []
+        splits_start_idxs = []
+        split_at_len = len(self.split_at)
+        cur_start_idx = 0
         cur_page = 1
         segments = windowed(elements, n=split_length, step=split_length - split_overlap)
+
         for seg in segments:
             current_units = [unit for unit in seg if unit is not None]
             txt = "".join(current_units)
+
             # check if length of current units is below split_threshold
             if len(current_units) < split_threshold and len(text_splits) > 0:
                 # concatenate the last split with the current one
                 text_splits[-1] += txt
+
             elif len(txt) > 0:
                 text_splits.append(txt)
                 splits_pages.append(cur_page)
+                splits_start_idxs.append(cur_start_idx)
+
             processed_units = current_units[: split_length - split_overlap]
+            cur_start_idx += len("".join(processed_units)) + split_at_len
+
             if self.split_by == "page":
                 num_page_breaks = len(processed_units)
             else:
                 num_page_breaks = sum(processed_unit.count("\f") for processed_unit in processed_units)
+
             cur_page += num_page_breaks
-        return text_splits, splits_pages
 
-    @staticmethod
-    def _create_docs_from_splits(text_splits: List[str], splits_pages: List[int], meta: Dict) -> List[Document]:
+        return text_splits, splits_pages, splits_start_idxs
+
+    def _create_docs_from_splits(
+        self, text_splits: List[str], splits_pages: List[int], splits_start_idxs: List[int], meta: Dict
+    ) -> List[Document]:
         """
         Creates Document objects from splits enriching them with page number and the metadata of the original document.
         """
         documents: List[Document] = []
 
-        for i, txt in enumerate(text_splits):
+        for i, (txt, split_idx) in enumerate(zip(text_splits, splits_start_idxs)):
             meta = deepcopy(meta)
             doc = Document(content=txt, meta=meta)
             doc.meta["page_number"] = splits_pages[i]
+            doc.meta["split_id"] = i
+            doc.meta["split_idx_start"] = split_idx
             documents.append(doc)
+
+            if self.split_overlap > 0:
+                doc.meta["_split_overlap"] = []
+                if i != 0:
+                    doc_start_idx = splits_start_idxs[i]
+                    previous_doc = documents[i - 1]
+                    previous_doc_start_idx = splits_start_idxs[i - 1]
+                    self._add_split_overlap_information(doc, doc_start_idx, previous_doc, previous_doc_start_idx)
```

Comment:
I would change this like this, what do you think?

```suggestion
            if self.split_overlap <= 0:
                continue

            doc.meta["_split_overlap"] = []

            if i == 0:
                continue

            doc_start_idx = splits_start_idxs[i]
            previous_doc = documents[i - 1]
            previous_doc_start_idx = splits_start_idxs[i - 1]
            self._add_split_overlap_information(doc, doc_start_idx, previous_doc, previous_doc_start_idx)

```

Reply:
I'm OK with it, guess it makes easier to read?

Reply:
Yeah, that's the idea. 😁

Reply:
got it, done 👍🏽